In [1]:
%%capture
%load_ext sql
%sql sqlite:///chinook.db

'Connected: None@chinook.db'

In [2]:
%%sql
SELECT
    name,
    type
FROM sqlite_master
WHERE type IN ("table","view");

Done.


[('album', 'table'),
 ('artist', 'table'),
 ('customer', 'table'),
 ('employee', 'table'),
 ('genre', 'table'),
 ('invoice', 'table'),
 ('invoice_line', 'table'),
 ('media_type', 'table'),
 ('playlist', 'table'),
 ('playlist_track', 'table'),
 ('track', 'table')]

In [3]:
%%sql
WITH usa_customers AS
    (SELECT *
       FROM customer
      WHERE country = "USA")

SELECT *
  FROM usa_customers
 LIMIT 3;

Done.


[(16, 'Frank', 'Harris', 'Google Inc.', '1600 Amphitheatre Parkway', 'Mountain View', 'CA', 'USA', '94043-1351', '+1 (650) 253-0000', '+1 (650) 253-0000', 'fharris@google.com', 4),
 (17, 'Jack', 'Smith', 'Microsoft Corporation', '1 Microsoft Way', 'Redmond', 'WA', 'USA', '98052-8300', '+1 (425) 882-8080', '+1 (425) 882-8081', 'jacksmith@microsoft.com', 5),
 (18, 'Michelle', 'Brooks', None, '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', '+1 (212) 221-3546', '+1 (212) 221-4679', 'michelleb@aol.com', 3)]

In [4]:
%%sql
WITH tracks_and_genre AS
(
SELECT il.invoice_id,
       g.name genre,
       SUM(il.quantity) number_of_tracks_sold,
       SUM(CAST(il.quantity AS float)) / 
            (SELECT SUM(quantity)
               FROM invoice_line) percentage_tracks_sold
  FROM invoice_line il
  LEFT JOIN track t
    ON il.track_id = t.track_id
  LEFT JOIN genre g
    ON t.genre_id = g.genre_id
 GROUP BY genre )

SELECT *
  FROM tracks_and_genre;

Done.


[(600, 'Alternative', 117, 0.024595333193188986),
 (609, 'Alternative & Punk', 492, 0.10342652932520496),
 (595, 'Blues', 124, 0.026066848854319948),
 (586, 'Classical', 47, 0.009880176581879336),
 (29, 'Drama', 1, 0.00021021652301870928),
 (610, 'Easy Listening', 74, 0.015556022703384485),
 (566, 'Electronica/Dance', 55, 0.01156190876602901),
 (558, 'Heavy Metal', 8, 0.0016817321841496743),
 (577, 'Hip Hop/Rap', 33, 0.006937145259617406),
 (613, 'Jazz', 121, 0.025436199285263823),
 (610, 'Latin', 167, 0.03510615934412445),
 (613, 'Metal', 619, 0.13012402774858103),
 (603, 'Pop', 63, 0.013243640950178683),
 (613, 'R&B/Soul', 159, 0.033424427159974776),
 (592, 'Reggae', 35, 0.007357578305654824),
 (614, 'Rock', 2635, 0.553920538154299),
 (441, 'Soundtrack', 5, 0.0010510826150935463),
 (383, 'TV Shows', 2, 0.00042043304603741857)]

In [5]:
%%sql

SELECT 
       g.name genre,
       SUM(il.quantity) number_of_tracks_sold,
       SUM(CAST(il.quantity AS float)) / 
            (SELECT SUM(il.quantity)
               FROM invoice_line il
               LEFT JOIN invoice i
                 ON il.invoice_id = i.invoice_id
               LEFT JOIN customer c
                 ON i.customer_id = c.customer_id
              WHERE c.country = "USA") percentage_tracks_sold
  FROM invoice_line il
  LEFT JOIN track t
    ON il.track_id = t.track_id
  LEFT JOIN genre g
    ON t.genre_id = g.genre_id
  LEFT JOIN invoice i
    ON il.invoice_id = i.invoice_id
  LEFT JOIN customer c
    ON i.customer_id = c.customer_id
 WHERE c.country = "USA"
 GROUP BY genre
 ORDER BY number_of_tracks_sold DESC
 LIMIT 10;

Done.


[('Rock', 561, 0.5337773549000951),
 ('Alternative & Punk', 130, 0.12369172216936251),
 ('Metal', 124, 0.11798287345385347),
 ('R&B/Soul', 53, 0.05042816365366318),
 ('Blues', 36, 0.03425309229305423),
 ('Alternative', 35, 0.03330161750713606),
 ('Latin', 22, 0.02093244529019981),
 ('Pop', 22, 0.02093244529019981),
 ('Hip Hop/Rap', 20, 0.019029495718363463),
 ('Jazz', 14, 0.013320647002854425)]

## Top Selling Genres in the USA

Rock is by far the best selling genre in the UK, with over 50% of all tracks sold.  The next top sellers are Alternative and Punk and Metal.  The top 3 best selling albums are also from Rock genre.

In [6]:
%%sql
SELECT at.name artist_name,
       al.title album,
       g.name genre,
       SUM(il.quantity) album_sales
  FROM track t
  LEFT JOIN album al
    ON t.album_id = al.album_id
  LEFT JOIN artist at
    ON al.artist_id = at.artist_id
  LEFT JOIN genre g
    ON t.genre_id = g.genre_id
  LEFT JOIN invoice_line il
    ON t.track_id = il.track_id
  LEFT JOIN invoice i
    ON i.invoice_id = il.invoice_id
  LEFT JOIN customer c
    ON i.customer_id = c.customer_id
 WHERE c.country = "USA"
 GROUP BY album
 ORDER BY album_sales DESC
 LIMIT 3;


Done.


[('Jimi Hendrix', 'Are You Experienced?', 'Rock', 28),
 ('Nirvana', 'From The Muddy Banks Of The Wishkah [live]', 'Rock', 28),
 ('The Doors', 'The Doors', 'Rock', 27)]

In [7]:
%%sql
SELECT e.first_name || " " || e.last_name employee_name,
       e.hire_date,
       ROUND(SUM(i.total), 2) total_sales,
       ROUND(AVG(i.total), 2) average_sale
  FROM employee e
 INNER JOIN customer c
    ON e.employee_id = c.support_rep_id
 INNER JOIN invoice i
    ON c.customer_id = i.customer_id
 WHERE title = "Sales Support Agent"
 GROUP BY employee_name

Done.


[('Jane Peacock', '2017-04-01 00:00:00', 1731.51, 8.17),
 ('Margaret Park', '2017-05-03 00:00:00', 1584.0, 7.4),
 ('Steve Johnson', '2017-10-17 00:00:00', 1393.92, 7.41)]

## Sales made by Sales Support Agents

There are three sales support agents, Jane Peacock makes the most sales at \$1731.51.  She also makes the largest sale at \$8.17.  Margaret Park makes \$1584 in total sales with average sale of \$7.4.  Steve Johnson makes a similar average sale to Margaret at \$7.41, but because Steve has been working 5 months less than Margaret his total sales are only \$1393.92.  Jane has been working the longest, which may account somewhat for her higher sales.

In [8]:
%%sql

WITH countries_other AS
    (SELECT CASE
               WHEN (
                      SELECT COUNT(*)
                        FROM customer 
                       WHERE country = c.country) = 1
                      THEN "Other"
                      ELSE c.country
                      END country,
               c.customer_id,
               i.*
      FROM customer c
      LEFT JOIN invoice i
        ON c.customer_id = i.customer_id)

SELECT country,
       COUNT(DISTINCT(customer_id)) number_of_customers,
       ROUND(SUM(total), 2) total_sales,
       ROUND(AVG(total), 2) avg_sale,
       ROUND(SUM(total) / COUNT(DISTINCT(customer_id))) avg_customer_total,
       CASE
          WHEN country = "Other" THEN 1
          ELSE 0
          END AS sort
  FROM countries_other
 GROUP BY country
 ORDER BY sort, number_of_customers DESC

Done.


[('USA', 13, 1040.49, 7.94, 80.0, 0),
 ('Canada', 8, 535.59, 7.05, 67.0, 0),
 ('Brazil', 5, 427.68, 7.01, 86.0, 0),
 ('France', 5, 389.07, 7.78, 78.0, 0),
 ('Germany', 4, 334.62, 8.16, 84.0, 0),
 ('United Kingdom', 3, 245.52, 8.77, 82.0, 0),
 ('Czech Republic', 2, 273.24, 9.11, 137.0, 0),
 ('India', 2, 183.15, 8.72, 92.0, 0),
 ('Portugal', 2, 185.13, 6.38, 93.0, 0),
 ('Other', 15, 1094.94, 7.45, 73.0, 1)]

## Country Customers Sales

The country with the most customers is the USA with almost twice as many as Canada, this is also reflected in total sales.  Some other countries have higher average sales or average customer total sales, like Czech Republic.  Countries with only 1 customer have all been placed in "Other" category, and together they have sales of 1094, slightly higher than the USA, but this is the combined total from 15 countries.

In [28]:
%%sql

WITH invoices_albums AS
    (SELECT a.album_id, 
            t.track_id album_track,
            il.invoice_id,
            il.track_id invoice_track
       FROM album a
       LEFT JOIN track t
         ON a.album_id = t.album_id
       LEFT JOIN invoice_line il
         ON t.track_id = il.track_id)
    
SELECT 

Done.


[(91, 1158, 1, 1158, 'True'),
 (91, 1159, 1, 1159, 'True'),
 (91, 1160, 1, 1160, 'True'),
 (91, 1161, 1, 1161, 'True'),
 (91, 1162, 1, 1162, 'True'),
 (91, 1163, 1, 1163, 'True'),
 (91, 1164, 1, 1164, 'True'),
 (91, 1165, 1, 1165, 'True'),
 (91, 1166, 1, 1166, 'True'),
 (91, 1167, 1, 1167, 'True'),
 (91, 1168, 1, 1168, 'True'),
 (91, 1169, 1, 1169, 'True'),
 (91, 1170, 1, 1170, 'True'),
 (91, 1171, 1, 1171, 'True'),
 (91, 1172, 1, 1172, 'True'),
 (91, 1173, 1, 1173, 'True'),
 (34, 392, 2, 392, 'True'),
 (20, 201, 2, 201, 'True'),
 (134, 1641, 2, 1641, 'True'),
 (39, 482, 2, 482, 'True'),
 (66, 819, 2, 819, 'True'),
 (73, 1119, 2, 1119, 'True'),
 (190, 2324, 2, 2324, 'True'),
 (215, 2651, 2, 2651, 'True'),
 (218, 2701, 2, 2701, 'True'),
 (322, 3476, 2, 3476, 'True'),
 (203, 2516, 3, 2516, 'True'),
 (214, 2646, 3, 2646, 'True'),
 (153, 1872, 4, 1872, 'True'),
 (58, 748, 4, 748, 'True'),
 (67, 829, 4, 829, 'True'),
 (146, 1778, 4, 1778, 'True'),
 (203, 2514, 4, 2514, 'True'),
 (207, 2560, 4, 2560, 'True'),
 (260, 3336, 4, 3336, 'True'),
 (314, 3448, 4, 3448, 'True'),
 (163, 1986, 5, 1986, 'True'),
 (163, 1987, 5, 1987, 'True'),
 (163, 1988, 5, 1988, 'True'),
 (163, 1989, 5, 1989, 'True'),
 (163, 1990, 5, 1990, 'True'),
 (163, 1991, 5, 1991, 'True'),
 (163, 1992, 5, 1992, 'True'),
 (163, 1993, 5, 1993, 'True'),
 (163, 1994, 5, 1994, 'True'),
 (163, 1995, 5, 1995, 'True'),
 (163, 1996, 5, 1996, 'True'),
 (163, 1997, 5, 1997, 'True'),
 (163, 1998, 5, 1998, 'True'),
 (163, 1999, 5, 1999, 'True'),
 (163, 2000, 5, 2000, 'True'),
 (163, 2001, 5, 2001, 'True'),
 (163, 2002, 5, 2002, 'True'),
 (5, 30, 6, 30, 'True'),
 (83, 1045, 6, 1045, 'True'),
 (6, 42, 7, 42, 'True'),
 (87, 1102, 7, 1102, 'True'),
 (55, 711, 7, 711, 'True'),
 (67, 834, 7, 834, 'True'),
 (106, 1343, 7, 1343, 'True'),
 (163, 1988, 7, 1988, 'True'),
 (211, 2596, 7, 2596, 'True'),
 (214, 2644, 7, 2644, 'True'),
 (239, 3024, 7, 3024, 'True'),
 (305, 3439, 7, 3439, 'True'),
 (322, 3477, 7, 3477, 'True'),
 (9, 81, 8, 81, 'True'),
 (185, 2263, 8, 2263, 'True'),
 (46, 574, 8, 574, 'True'),
 (118, 1455, 8, 1455, 'True'),
 (125, 1560, 8, 1560, 'True'),
 (144, 1752, 8, 1752, 'True'),
 (145, 1758, 8, 1758, 'True'),
 (183, 2231, 8, 2231, 'True'),
 (203, 2509, 8, 2509, 'True'),
 (233, 2946, 8, 2946, 'True'),
 (271, 3399, 9, 3399, 'True'),
 (20, 196, 9, 196, 'True'),
 (21, 206, 9, 206, 'True'),
 (79, 993, 9, 993, 'True'),
 (83, 1041, 9, 1041, 'True'),
 (101, 1279, 9, 1279, 'True'),
 (162, 1983, 9, 1983, 'True'),
 (173, 2098, 9, 2098, 'True'),
 (321, 3456, 9, 3456, 'True'),
 (215, 2663, 10, 2663, 'True'),
 (257, 3290, 10, 3290, 'True'),
 (185, 2257, 11, 2257, 'True'),
 (89, 1138, 11, 1138, 'True'),
 (197, 2420, 11, 2420, 'True'),
 (49, 610, 11, 610, 'True'),
 (115, 1419, 11, 1419, 'True'),
 (181, 2203, 11, 2203, 'True'),
 (195, 2391, 11, 2391, 'True'),
 (211, 2597, 11, 2597, 'True'),
 (214, 2643, 11, 2643, 'True'),
 (243, 3072, 11, 3072, 'True'),
 (318, 3452, 11, 3452, 'True'),
 (10, 92, 12, 92, 'True'),
 (54, 690, 12, 690, 'True'),
 (118, 1463, 12, 1463, 'True'),
 (215, 2663, 12, 2663, 'True'),
 (206, 2553, 13, 2553, 'True'),
 (42, 541, 14, 541, 'True'),
 (65, 807, 15, 807, 'True'),
 (120, 1489, 15, 1489, 'True'),
 (241, 3050, 15, 3050, 'True'),
 (260, 3336, 15, 3336, 'True'),
 (4, 16, 16, 16, 'True'),
 (4, 21, 16, 21, 'True'),
 (5, 24, 16, 24, 'True'),
 (9, 83, 16, 83, 'True'),
 (10, 93, 16, 93, 'True'),
 (70, 875, 16, 875, 'True'),
 (145, 1756, 16, 1756, 'True'),
 (178, 2160, 16, 2160, 'True'),
 (321, 3460, 16, 3460, 'True'),
 (7, 55, 17, 55, 'True'),
 (7, 59, 17, 59, 'True'),
 (131, 1610, 17, 1610, 'True'),
 (185, 2270, 17, 2270, 'True'),
 (115, 1419, 17, 1419, 'True'),
 (162, 1973, 17, 1973, 'True'),
 (179, 2171, 17, 2171, 'True'),
 (195, 2400, 17, 2400, 'True'),
 (203, 2522, 17, 2522, 'True'),
 (206, 2553, 17, 2553, 'True'),
 (207, 2564, 17, 2564, 'True'),
 (127, 1583, 18, 1583, 'True'),
 (82, 1027, 18, 1027, 'True'),
 (141, 3142, 18, 3142, 'True'),
 (144, 1752, 18, 1752, 'True')

In [67]:
%%sql
SELECT a.name artist,
       COUNT(DISTINCT(plt.playlist_id)) number_of_playlists
  FROM playlist_track plt
 INNER JOIN track t
    ON plt.track_id = t.track_id
 INNER JOIN album al
    ON t.album_id = al.album_id
 INNER JOIN artist a
    ON al.artist_id = a.artist_id
 GROUP BY artist
 ORDER BY number_of_playlists DESC
 LIMIT 5;
 


Done.


[('Eugene Ormandy', 7),
 ('Academy of St. Martin in the Fields & Sir Neville Marriner', 6),
 ('Berliner Philharmoniker & Herbert Von Karajan', 6),
 ('English Concert & Trevor Pinnock', 6),
 ("The King's Singers", 6)]

## Artists in Playlists

The artist in the most playlists is Eugene Ormandy with 7.

In [121]:
%%sql

SELECT CASE
           WHEN il.track_id IS NULL THEN "Not purchased"
           ELSE "Purchased"
           END AS was_purchased,
        COUNT(DISTINCT(t.track_id)) count,
        CAST(COUNT(DISTINCT(t.track_id)) as float) / (SELECT COUNT(*)
           FROM track) percentage
  FROM track t
  LEFT JOIN invoice_line il
    ON t.track_id = il.track_id
 GROUP BY was_purchased


Done.


[('Not purchased', 1697, 0.4844419069369112),
 ('Purchased', 1806, 0.5155580930630888)]

## Purchased Tracks

Roughly half of all tracks available have been purchased.

In [125]:
%%sql

SELECT mt.name,
       COUNT(*),
       ROUND(CAST(COUNT(*) AS float) / (SELECT COUNT(*)
                                    FROM track), 2) percentage
  FROM track t
 INNER JOIN media_type mt
    ON t.media_type_id = mt.media_type_id
 GROUP BY mt.name
 

Done.


[('AAC audio file', 11, 0.0),
 ('MPEG audio file', 3034, 0.87),
 ('Protected AAC audio file', 237, 0.07),
 ('Protected MPEG-4 video file', 214, 0.06),
 ('Purchased AAC audio file', 7, 0.0)]

In [140]:
%%sql

SELECT COUNT(*) number_tracks_sold,
        CASE
          WHEN mt.name LIKE "%Protected%" THEN "Protected"
          ELSE "Unprotected"
          END AS protection,
       ROUND(CAST(COUNT(*) AS float) / (SELECT COUNT(*)
                                    FROM invoice_line), 2) percentage
  FROM invoice_line il
  LEFT JOIN track t
    ON il.track_id = t.track_id
  LEFT JOIN media_type mt
    ON t.media_type_id = mt.media_type_id
 GROUP BY protection

Done.


[(442, 'Protected', 0.09), (4315, 'Unprotected', 0.91)]

## Protected VS Unprotected

Protected music files make up about 13% of all tracks available in the music store, less than 10% of track sales.  Protection is not likely to affect popularity.